src/libinjector/linux/methods/linux_shellcode.c

```diff
@@ -0,0 +1,164 @@
+#include "linux_shellcode.h"
+#include "linux_debug.h"
+#include "linux_syscalls.h"
+
+bool setup_mmap_trap(drakvuf_t drakvuf, drakvuf_trap_info_t* info);
+bool write_shellcode_to_mmap_location(drakvuf_t drakvuf, drakvuf_trap_info_t* info);
+
+event_response_t handle_shellcode(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+
+    injector_t injector = (injector_t)info->trap->data;
+
+    event_response_t event;
+
+    switch (injector->step)
+    {
+        case STEP1: // Finds vdso and sets up mmap
+        {
+            memcpy(&injector->saved_regs, info->regs, sizeof(x86_registers_t));
+
+            addr_t vdso = find_vdso(drakvuf, info);
+            addr_t syscall_addr = find_syscall(drakvuf, info, vdso);
+
+            setup_post_syscall_trap(drakvuf, info, syscall_addr);
+            // don't remove the initial trap
+            // it is used for cleanup after restoring registers
+
+            if (!setup_mmap_syscall(injector, info->regs, 4096))
+                PRINT_DEBUG("Failed to setup mmap syscall");
+
+            info->regs->rip = syscall_addr;
+            info->regs->rax = injector->syscall_no;
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+
+            break;
+        }
+        case STEP2: // setup shellcode
+        {
+            PRINT_DEBUG("mmap location is: %lx\n", info->regs->rax);
+
+            // save it for future use
+            injector->virtual_memory_addr = info->regs->rax;
+
+            if (write_shellcode_to_mmap_location(drakvuf, info))
+            {
+                setup_mmap_trap(drakvuf, info);
+                info->regs->rip = injector->virtual_memory_addr;
+            }
+
+            free_bp_trap(drakvuf, injector, info->trap);
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP3: //since mmap starting location is trapped, the first one will be this
+        {
+            PRINT_DEBUG("Shellcode begin\n");
+            save_rip_for_ret(drakvuf, info->regs);
+
+            // rsp is being updated
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP4: // shellcode should've executed and we will be returned to the same trap as ret is appended in the end
+        {
+            PRINT_DEBUG("Shellcode end\n");
+            free_bp_trap(drakvuf, injector, info->trap);
+
+            // restore regs
+            memcpy(info->regs, &injector->saved_regs, sizeof(x86_registers_t));
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP5: // cleanup
```

Comment:
Just asking - why can't this step be in STEP4? I can see you already restore process registers there.

Reply:
umm, I wasn't sure about `drakvuf_interrupt`, I was guessing that if I interrupt and also return `VMI_EVENT_RESPONSE_SET_REGISTERS`, it thought that the registers won't be set and interrupt will be given the priority, I'll try combining those steps

Reply:
I tried it, somehow when I am trying to clear everything in `STEP4` the shellcode is being run twice somehow, I don't know why. Can we keep it separate only if that's not an issue?

Reply:
Ok